include/circt/Dialect/FIRRTL/FIRRTLStatements.td

```diff
@@ -131,12 +131,16 @@ def FPrintFOp : FIRRTLOp<"fprintf"> {
   }];
 
   let arguments = (ins ClockType:$clock, UInt1Type:$cond, StrAttr:$outputFile,
+                       // Use FStringType instead of PrintfOperandType to reject hardware types.
+                       Variadic<FStringType>:$outputFileSubstitutions,
```

Comment:
This is a good use of `FStringType` until we want to (if we want to) expand it later.

Reply:
This is replaced with PrintfOperand type as it's necessary to handle as well. A side effect of PrintfOperand not being buildable type is that assembly format got too hard to keep the consistency with other operations. So I end up implementing custom assembly format.